applications/CompressiblePotentialFlowApplication/custom_utilities/potential_flow_utilities.cpp

```diff
@@ -680,6 +692,116 @@ double ComputeUpwindedDensity(
+
+    KRATOS_ERROR_IF(free_stream_mach < std::numeric_limits<double>::epsilon())
+        << "ComputeDensityDerivativeWRTVelocitySquared: free stream mach number must be larger than zero." << std::endl;
+
+    const double local_velocity_squared = ComputeVelocityMagnitude<Dim, NumNodes>(localMachNumberSquared, rCurrentProcessInfo);
+
+    // ratio of speed of sound to free stream speed of sound
+    const double speed_of_sound_ratio = ComputeSquaredSpeedofSoundFactor<Dim, NumNodes>(local_velocity_squared, rCurrentProcessInfo);
+
+    const double free_stream_values_const = -0.5 * free_stream_density * std::pow(free_stream_mach,2.0) / inner_prod(free_stream_velocity, free_stream_velocity);
+    const double speed_of_sound_power = (2.0 - heat_capacity_ratio) / (heat_capacity_ratio - 1.0);
+
+    KRATOS_ERROR_IF((heat_capacity_ratio - 1.0) < std::numeric_limits<double>::epsilon())
+        << "ComputeDensityDerivativeWRTVelocitySquared: heat capacity ratio must not be 1.0." << std::endl;
+
+    return free_stream_values_const * std::pow(speed_of_sound_ratio, speed_of_sound_power);
+}
+
+template <int Dim, int NumNodes>
+double ComputeUpwindedDensityDerivativeWRTVelocitySquaredSupersonicAccelerating(
+    const array_1d<double, Dim>& rCurrentVelocity, 
+    const double currentMachNumberSquared,
+    const double upwindMachNumberSquared, 
+    const ProcessInfo& rCurrentProcessInfo)
+{
+    // Following Fully Simulataneous Coupling of the Full Potential Equation
+    //           and the Integral Boundary Layer Equations in Three Dimensions
+    //           by Brian Nishida (1996), Section A.2.6
+    const double upwind_factor = ComputeUpwindFactor<Dim, NumNodes>(currentMachNumberSquared, rCurrentProcessInfo);
+
+    const double upwind_factor_derivative = ComputeUpwindFactorDerivativeWRTVelocitySquared<Dim, NumNodes>(rCurrentVelocity, rCurrentProcessInfo);
+
+    const double Drho_Dq2 = ComputeDensityDerivativeWRTVelocitySquared<Dim, NumNodes>(currentMachNumberSquared, rCurrentProcessInfo);
+
+    const double current_density = ComputeDensity<Dim, NumNodes>(currentMachNumberSquared, rCurrentProcessInfo);
+    const double upwind_density = ComputeDensity<Dim, NumNodes>(upwindMachNumberSquared, rCurrentProcessInfo);
+
+    return Drho_Dq2 * (1.0 - upwind_factor) - upwind_factor_derivative * (current_density - upwind_density);
+}
+
+template <int Dim, int NumNodes>
+double ComputeUpwindedDensityDerivativeWRTVelocitySquaredSupersonicDeaccelerating(
+    const double currentMachNumberSquared,
+    const double upwindMachSquared,
+    const ProcessInfo& rCurrentProcessInfo)
+{
+    // Following Fully Simulataneous Coupling of the Full Potential Equation
+    //           and the Integral Boundary Layer Equations in Three Dimensions
+    //           by Brian Nishida (1996), Section A.2.6    
+    // const double current_mach_sq = ComputeLocalMachNumberSquared<Dim, NumNodes>(rCurrentVelocity, rCurrentProcessInfo);
```

Comment:
We usually try not to commit unused commented lines of code into master, do you need this one for later?
```suggestion
```